### PR TITLE
Support _ in spark conf of integration tests

### LINKED ISSUE #6351: [BUG] Implement escape characters for spark property encoding in PYSP_TEST  env variables
> It works for non-xdist execution, but not work for xdist execution.
export TEST_PARALLEL=2
Then run, will get: ClassNotFoundException

Thanks for pointing this out. I was using the command from https://github.com/NVIDIA/spark-rapids/blob/49e0a85e930ec98635269a703069f631f6f7b419/jenkins/spark-tests.sh#L201-L207 as a reference.

The issue is the xdist call path cannot support iceberg tests at all.
`SPARK_SUBMIT_FLAGS` is only respected in the non-xdist call path. The correct way to launch it with xdist would be:
```bash
export TEST_PARALLEL=2
export SPARK_HOME=~/dist/spark-3.2.2-bin-hadoop3.2
export NUM_LOCAL_EXECS="1"
export PYSP_TEST_spark_jars_packages="org.apache.iceberg:iceberg-spark-runtime-3.2_2.12:0.13.2"
export PYSP_TEST_spark_sql_extensions="org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions"
dexport PYSP_TEST_spark_sql_catalog_spark_catalog="org.apache.iceberg.spark.SparkSessionCatalog"
export PYSP_TEST_spark_sql_catalog_spark_catalog_type="hadoop"
export PYSP_TEST_spark_sql_catalog_spark_catalog_warehouse="$(mktemp -d -p /tmp spark-warehouse-XXXXXX)"
./integration_tests/run_pyspark_from_build.sh -m iceberg --iceberg -k 'test_iceberg_read_partition_key[PERFILE-Long]'
```

However, we currently have no way to escape characters in spark properties as is required for '_' in `spark_catalog`.

We should work on this in a separate issue.

_Originally posted by @gerashegalov in https://github.com/NVIDIA/spark-rapids/issues/6342#issuecomment-1218310299_

### PR DESCRIPTION
**Pull request by gerashegalov**
Fixes #6351
- Allow escaping _  in Spark conf by duplicating _ to allow running Iceberg/ pytests with xdist

#### Additional refactoring and fixes
- Remove a stale comment from spark-tests.sh
- Allow overriding PYSP_TEST_spark_rapids_memory_gpu_allocSize to be able to run multiple Spark apps with multiple executors per pytest session with TEST_PARALLEL >= 2 on a small GPU 
- Add support for --packages and --jars in xdist using findspark
  - Fix xdist failing to init findspark when pyspark already on PYTHONPATH
  - Enforce findspark usage with xdist
    - don't start Spark and allocate GPU resources for xdist master
    - remove direct imports of pyspark from conftest to make sure it happens after session start

Will file a follow-up issue to convert non-xdist tests relying on `SPARK_SUBMIT_ARGS` for external sources such as Delta and Iceberg to xdist . The following works locally:

```bash
export TEST_PARALLEL=2 
export SPARK_HOME=~/dist/spark-3.2.2-bin-hadoop3.2
export PYSP_TEST_spark_master="spark://$(hostname):7077"
export PYSP_TEST_spark_executor_cores="2"
export PYSP_TEST_spark_cores_max="4"
export PYSP_TEST_spark_jars_packages="org.apache.iceberg:iceberg-spark-runtime-3.2_2.12:0.13.2"
export PYSP_TEST_spark_sql_extensions="org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions"
export PYSP_TEST_spark_sql_catalog_spark__catalog="org.apache.iceberg.spark.SparkSessionCatalog"
export PYSP_TEST_spark_sql_catalog_spark__catalog_type="hadoop"
export PYSP_TEST_spark_sql_catalog_spark__catalog_warehouse="$(mktemp -d -p /tmp spark-warehouse-XXXXXX)"
export PYSP_TEST_spark_rapids_memory_gpu_allocSize='512m'
./integration_tests/run_pyspark_from_build.sh -s -m iceberg --iceberg
 ```

Signed-off-by: Gera Shegalov <gera@apache.org>

### CODE REVIEW

**Comment by gerashegalov** [comment]
build

File: integration_tests/README.md [hunk 1]
```diff
@@ -37,20 +37,20 @@ It is recommended that you use `pyenv` to manage Python installations.
 - Follow instructions to use the right method of installation described
   [here](https://github.com/pyenv/pyenv#installation)
 - Verify that `pyenv` is set correctly
-
```
Comment (revans2) [inline integration_tests/README.md:40]: Why is this file being changed when it does not appear to be related to anything else in the PR?

File: integration_tests/run_pyspark_from_build.sh [hunk 2]
```diff
@@ -258,19 +258,25 @@ else
 
     # If you want to change the amount of GPU memory allocated you have to change it here
     # and where TEST_PARALLEL is calculated
-    export PYSP_TEST_spark_rapids_memory_gpu_allocSize='1536m'
+    export PYSP_TEST_spark_rapids_memory_gpu_allocSize=${PYSP_TEST_spark_rapids_memory_gpu_allocSize:-'1536m'}
```
Comment (revans2) [inline integration_tests/run_pyspark_from_build.sh:261]: Why do we need this change?  The point of this was to set an amount that was always consistent no matter what GPU we run on.  

As a separate note the calculations for TEST_PARALLEL were not changed to take this into account.

https://github.com/NVIDIA/spark-rapids/blob/814732c13c3621698d9b29deff7c70b7cd263b62/integration_tests/run_pyspark_from_build.sh#L124

If you really want this to be configurable then we need to either error out if TEST_PARALLEL is not set and gpu_allocSize is, or we need to make sure that the TEST_PARALLEL calculation take the allocSize into account properly.

I personally see no reason to make this configurable.

File: jenkins/spark-tests.sh [hunk 3]
```diff
@@ -195,9 +195,6 @@ run_iceberg_tests() {
 
   # Iceberg does not support Spark 3.3+ yet
   if [[ "$ICEBERG_SPARK_VER" < "3.3" ]]; then
-    # Classloader config is here to work around classloader issues with
```
Comment (revans2) [inline jenkins/spark-tests.sh:198]: Why is this removed as a part of this PR?

**Review by revans2 — COMMENTED** [review]
Most of the issues I have are nits. Combining a lot of small things together into a single PR is not ideal for me, but I am okay with it. Just confused because it was not reflected in the name of the PR at all.  I had to read the description to know what was happening.

The one I really care about is setting the memory size for the test.

File: integration_tests/README.md (hunk 1 above)
Comment (gerashegalov) [reply integration_tests/README.md:40]: To explain how the user should translate a Spark config key with an underscore to the environment variable for integration tests

https://github.com/NVIDIA/spark-rapids/pull/6358/files#diff-2fd87c8f6ba9eceb00af05de7edefb7b5e70706cd27bebf9d923da293af4476bR232-R235

The rest is my editor trimming trailing white spaces . You can hit "Hide whitespace"

File: integration_tests/run_pyspark_from_build.sh (hunk 2 above)
Comment (gerashegalov) [reply integration_tests/run_pyspark_from_build.sh:261]: I did not try to make the magic formula work for my system. As is, my system freezes when I configure a fully distributed setup as in the description: 2 pytest apps with 2 executors each. On my notebook GPU I typically have 6G available and have to modify this variable in the file to something smaller like 512m.

File: jenkins/spark-tests.sh (hunk 3 above)
Comment (gerashegalov) [reply jenkins/spark-tests.sh:198]: it should have been removed in #5646 because the comment refers to force.caller.classloader=false that no longer exists.

**Comment by gerashegalov** [comment]
Can't repro the CI failure locally, retrying

**Comment by gerashegalov** [comment]
build

**Comment by gerashegalov** [comment]
Since it's a very basic ClassNofFound for SQLPlugin class it may be related to a potential findspark version discrepancy.

**Comment by gerashegalov** [comment]
build

**Comment by gerashegalov** [comment]
The reason there is a discrepancy with the CI is that CI explicitly puts pyspark on the PYTHONPATH
```
2022-08-18T12:53:56.6273334Z [2022-08-18T12:46:09.218Z] + export PYTHONPATH=/home/jenkins/agent/workspace/jenkins-rapids_premerge-github-5402-ci-2/.download/spark-3.1.1-bin-hadoop3.2/python:/home/jenkins/agent/workspace/jenkins-rapids_premerge-github-5402-ci-2/.download/spark-3.1.1-bin-hadoop3.2/python/pyspark/:/home/jenkins/agent/workspace/jenkins-rapids_premerge-github-5402-ci-2/.download/spark-3.1.1-bin-hadoop3.2/python/lib/py4j-0.10.9-src.zip
```

Thus the code that relies on pyspark not being found to fallback on findspark initialization does not kick in

**Comment by gerashegalov** [comment]
build

**Comment by gerashegalov** [comment]
build

**Comment by gerashegalov** [comment]
build

**Comment by gerashegalov** [comment]
build

**Comment by gerashegalov** [comment]
build

**Comment by gerashegalov** [comment]
build

File: integration_tests/run_pyspark_from_build.sh (hunk 2 above)
Comment (revans2) [reply integration_tests/run_pyspark_from_build.sh:261]: Okay I understand the issue now. I think this is fine. Dropping the size is fine. There are/were some tests that needed 400+ MiB of GPU memory to work, and CUDF needs 500+ MiB so I rounded them both up to 1.5 GiB, but if you can manage with a lot less it is fine.  Just add a comment here so that it is clear it should not be used in CI tests and making a change is at your own risk.

**Review by revans2 — APPROVED** [review]
Thanks for explaining the changes.  This is fine. Just would like to see a comment on the one thing that is still left, but it is not critical.

File: integration_tests/run_pyspark_from_build.sh (hunk 2 above)
Comment (gerashegalov) [reply integration_tests/run_pyspark_from_build.sh:268]: I added a WARNING message to the script

**Comment by gerashegalov** [comment]
build

**Review by revans2 — APPROVED** [review]
